src/render/image.rs

```diff
@@ -0,0 +1,276 @@
+use std::{
+    cell::RefCell,
+    fs::File,
+    io::Read,
+    rc::Rc,
+    sync::atomic::{AtomicUsize, Ordering},
+    time::{Duration, Instant},
+};
+
+use apex_hardware::FrameBuffer;
+use embedded_graphics::{
+    image::{Image, ImageRaw},
+    pixelcolor::BinaryColor,
+    prelude::Point,
+    Drawable,
+};
+use image::{AnimationDecoder, DynamicImage};
+
+static GIF_MISSING: &[u8] = include_bytes!("./../../assets/gif_missing.gif");
+static DISPLAY_HEIGHT: i32 = 40;
+static DISPLAY_WIDTH: i32 = 128;
+
+pub struct ImageRenderer {
+    stop: Point,
+    origin: Point,
+    decoded_frames: Vec<Vec<u8>>,
+    current_frame: AtomicUsize,
+    delays: Vec<u16>,
+    time_frame_last_update: Rc<RefCell<Instant>>,
+}
+
+impl ImageRenderer {
+    pub fn calculate_median_color_value(
+        image: &image::ImageBuffer<image::Rgba<u8>, Vec<u8>>,
+        image_height: i32,
+        image_width: i32,
+    ) -> u8 {
+        //NOTE we're using the median to determine wether the pixel should be black or
+        // white
+
+        let mut colors = (0..=255).into_iter().map(|_| 0).collect::<Vec<u32>>();
+        let mut num_pixels_alpha = 0;
+
+        let height = image.height();
+        let width = image.width();
+
+        for y in 0..image_height {
+            //if y is outside of the gif width
+            if y >= height as i32 {
+                continue;
+            }
+
+            //if y is outside of the screen
+            if y >= DISPLAY_HEIGHT {
+                continue;
+            }
+            for x in 0..image_width {
+                //if x is outside of the gif width
+                if x >= width as i32 {
+                    continue;
+                }
+
+                //if x is outside of the screen
+                if x >= DISPLAY_WIDTH {
+                    continue;
+                }
+
+                let pixel = image.get_pixel(x as u32, y as u32);
+
+                let avg_pixel_value =
+                    ((u32::from(pixel[0]) + u32::from(pixel[1]) + u32::from(pixel[2])) / 3)
+                        as usize;
+
+                //the value is multiplied by the alpha (a) of said pixel
+                //the more the pixel is transparent, the less the pixel has an importance
+                colors[avg_pixel_value] += u32::from(pixel[3]) / 255;
```

Comment:
This is a bad way to do it now that I'm looking back, we should only /255 after we did every color, because since it's not a float it will either be a or 0 not a range like I intended it to be